mdn/kuma
# record subscriptions

Fixes #6579
Fixes #6684

- [x] Needs test for new feature in `api/v1/whoami`
- [x] Add new Stripe webhook for cancellations
- [x] Explain, in the PR, why certain changes were made (e.g. refactoring of checks, breaking up and moving `retrieve_stripe_subscription_info`, why the cancel method is a generator)


I really don't like some of the APIs in `payments/utils.py` that were about payments. For example, when you press to cancel your monthly subscription(s), it deletes all subscriptions you have. It'd be more sensible if it would list each subscription on the `payments/recurring/monthly` and there's a cancel for each one. 
Another thing is that in payments/recurring/management there's no information about each subscription so you can't sync-check the `UserSubscriptions`. Possibly joint with the point above. 
Also, I get nervous when I see all the unit tests for `payments/utils.py`, with heavy mocking, that probably shouldn't be tests at all. Instead, there should just be things in `payments/tests/test_views.py`

## How to test

Check out the branch and `docker-compose exec web ./manage.py migrate`
In one browser tab, go to http://localhost.org:8000/admin/users/usersubscription/
In another browser tab, go to http://localhost.org:8000/api/v1/whoami (note the `is_subscriber`)
In another tab go to `http://wiki.localhost.org:8000/en-US/profiles/$username/edit` and if you had a subscription already, go back to http://localhost.org:8000/admin/users/usersubscription/ and http://localhost.org:8000/api/v1/whoami and refresh. 
If you cancel your subscription, reload the above-mentioned pages and observe what happens. 
**Note** If you cancel your subscription and set up a new one, this'll be a *different* subscription so you'll get two records like this:
<img width="551" alt="Screen Shot 2020-03-16 at 11 29 51 AM" src="https://user-images.githubusercontent.com/26739/76773837-7cc1c980-6779-11ea-98f7-09be4a8989df.png">
